src/model/openthaigpt_pretraining_model/llama/model.py

```diff
@@ -1,14 +1,215 @@
+    dim: int = 512
+    n_layers: int = 8
+    n_heads: int = 8
+    vocab_size: int = -1  # defined later by tokenizer
+    multiple_of: int = 256  # make SwiGLU hidden layer size multiple of large power of 2
+    norm_eps: float = 1e-5
+
+    max_batch_size: int = 32
+    max_seq_len: int = 2048
+
+    attention_mode: str = ORIGIN_ATTENTION_MODE  # pytorch, xformer
+
+
+class Attention(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+
+        self.head_dim = args.dim // args.n_heads
+
+        self.n_local_heads = args.n_heads
+
+        self.wq = nn.Linear(
+            args.dim,
+            args.n_heads * self.head_dim,
+            bias=False,
+        )
+        self.wk = nn.Linear(
+            args.dim,
+            args.n_heads * self.head_dim,
+            bias=False,
+        )
+        self.wv = nn.Linear(
+            args.dim,
+            args.n_heads * self.head_dim,
+            bias=False,
+        )
+        self.wo = nn.Linear(
+            args.n_heads * self.head_dim,
+            args.dim,
+            bias=False,
+        )
+
+        self.cache_k = torch.zeros(
+            (args.max_batch_size, args.max_seq_len, self.n_local_heads, self.head_dim)
+        )
+        self.cache_v = torch.zeros(
+            (args.max_batch_size, args.max_seq_len, self.n_local_heads, self.head_dim)
+        )
+
+        if not (
+            args.attention_mode == PYTORCH_ATTENTION_MODE
+            or args.attention_mode == XFORMER_ATTENTION_MODE
+            or args.attention_mode == ORIGIN_ATTENTION_MODE
+        ):
+            raise KeyError(
+                f'attention mode must be "{ORIGIN_ATTENTION_MODE}", "{XFORMER_ATTENTION_MODE}" or "{PYTORCH_ATTENTION_MODE}"'  # noqa: E501
+            )
+
+        self.mode = args.attention_mode
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        start_pos: int,
+        freqs_cis: torch.Tensor,
+        mask: Optional[torch.Tensor],
+    ):
+        bsz, seqlen, _ = x.shape
+        xq, xk, xv = self.wq(x), self.wk(x), self.wv(x)
+
+        xq = xq.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+        xk = xk.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+        xv = xv.view(bsz, seqlen, self.n_local_heads, self.head_dim)
+
+        xq, xk = apply_rotary_emb(xq, xk, freqs_cis=freqs_cis)
+
+        self.cache_k = self.cache_k.to(xq)
+        self.cache_v = self.cache_v.to(xq)
+
+        self.cache_k[:bsz, start_pos : start_pos + seqlen] = xk  # noqa: E203
+        self.cache_v[:bsz, start_pos : start_pos + seqlen] = xv  # noqa: E203
+
+        keys = self.cache_k[:bsz, : start_pos + seqlen]
+        values = self.cache_v[:bsz, : start_pos + seqlen]
+
+        xq = xq.transpose(1, 2)
+        keys = keys.transpose(1, 2)
+        values = values.transpose(1, 2)
+
+        if self.mode == PYTORCH_ATTENTION_MODE:
+            output = torch_attention(xq, keys, values, mask)
+        elif self.mode == XFORMER_ATTENTION_MODE:

```

Comment:
there is another usecase of xformer attention mask **(and it should be faster!)**. It is called CasualBlockDiagonalMask 
please check if the input is a list of tensors. If yes, use `CasualBlockDiagonalMask` instead of `LowerTriangularMaskWithTensorBias`

Example of how to use `CasualBlockDiagonalMask`
```python
import torch
from xformers.ops import fmha

K = 16
dtype = torch.float16
device = "cuda"
list_x = [
    torch.randn([1, 3, 1, K], dtype=dtype, device=device),
    torch.randn([1, 6, 1, K], dtype=dtype, device=device),
    torch.randn([1, 2, 1, K], dtype=dtype, device=device),
]
attn_bias, x = fmha.BlockDiagonalMask.from_tensor_list(list_x)
linear = torch.nn.Linear(K, K * 3).to(device=device, dtype=dtype)

q, k, v = linear(x).reshape([1, -1, 1, 3, K]).unbind(-2)
out = fmha.memory_efficient_attention(q, k, v, attn_bias=attn_bias)
list_out = attn_bias.split(out)
print(list_out[0].shape)  # [1, 3, 1, K]
assert tuple(list_out[0].shape) == (1, 3, 1, K)
```
source: https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.fmha.attn_bias.BlockDiagonalMask.from_tensor_list

Reply:
input is only `torch.Tensor` can not be `list` of `torch.Tensor`